File: Java-CodeServer/src/main/resources/js/prompto/utils/CodeWriter.js
var Context = require("../runtime/Context").Context;

function Indenter() {
    this.value = "";
    this.indents = "";
    this.isStartOfLine = true;
    return this;
}

Indenter.prototype.appendTabsIfRequired = function(s) {
    if(this.isStartOfLine) {
        this.value += this.indents;
    }
    this.isStartOfLine = s.charAt(s.length-1)=='\n';
};

Indenter.prototype.append = function(s) {
    this.value += s;
};

Indenter.prototype.trimLast = function(count) {
    this.value = this.value.substring(0, this.value.length - count);
};

Indenter.prototype.indent = function() {
    this.indents += '\t';
};

Indenter.prototype.dedent = function() {
    if(this.indents.length==0) {
        throw new Exception("Illegal dedent!");
    }
    this.indents = this.indents.slice(1);
};


function CodeWriter(dialect, context, indenter) {
    this.dialect = dialect;
    this.context = context || Context.newGlobalContext();
    this.indenter = indenter || new Indenter();
    return this;
}

CodeWriter.prototype.isGlobalContext = function() {
    return this.context.isGlobalContext();
};

CodeWriter.prototype.append = function(s) {
    this.indenter.appendTabsIfRequired(s);
    this.indenter.append(s);
};

CodeWriter.prototype.toString = function() {
    return this.indenter.value;
};

CodeWriter.prototype.trimLast = function(count) {
    this.indenter.trimLast(count);
};

CodeWriter.prototype.indent = function() {
    this.indenter.indent();
};

CodeWriter.prototype.dedent = function() {
    this.indenter.dedent();
};

CodeWriter.prototype.newLine = function() {
    this.append('\n');
};

CodeWriter.prototype.newLocalWriter = function() {
    return new CodeWriter(this.dialect, this.context.newLocalContext(), this.indenter);
};

CodeWriter.prototype.newInstanceWriter = function(type) {
    return new CodeWriter(this.dialect, this.context.newInstanceContext(null, type), this.indenter);
};

CodeWriter.prototype.newMemberWriter = function() {
    var context = this.context.newLocalContext ();
    context.parent = this.context;
    return new CodeWriter (this.dialect, context, this.indenter);
};

CodeWriter.prototype.toDialect = function(o) {
    this.dialect.toDialect(this, o);
};

exports.CodeWriter = CodeWriter

File: Core/src/main/java/prompto/utils/CodeWriter.java
package prompto.utils;

import prompto.grammar.Identifier;
import prompto.parser.Dialect;
import prompto.runtime.Context;
import prompto.type.CategoryType;

public class CodeWriter {
	
	class Indenter {
		
		public String indents = "";
		public boolean isStartOfLine = true;
		
		public void appendTabsIfRequired(String s) {
			appendTabsIfRequired(s.charAt(s.length()-1));
		}

		public void appendTabsIfRequired(char c) {
			if(isStartOfLine)
				sb.append(indents);
			isStartOfLine = c=='\n';
		}
		
		public void indent() {
			indents += '\t';
		}

		public void dedent() {
			if(indents.isEmpty())
				throw new RuntimeException("Illegal dedent!");
			indents = indents.substring(1);
		}
	}
	
	Dialect dialect;
	Context context;
	StringBuilder sb;
	Indenter indenter;
	int escapeMode;
	
	public CodeWriter(Dialect dialect, Context context) {
		this(dialect, context, 0);
	}
	
	public CodeWriter(Dialect dialect, Context context, int escapeMode) {
		this.dialect = dialect;
		this.context = context;
		this.escapeMode = escapeMode;
		this.sb = new StringBuilder();
		indenter = new Indenter();
	}
	
	public CodeWriter(Dialect dialect, Context context, int escapeMode, StringBuilder sb, Indenter indenter) {
		this.dialect = dialect;
		this.context = context;
		this.escapeMode = escapeMode;
		this.sb = sb;
		this.indenter = indenter;
	}

	public Context getContext() {
		return context;
	}
	
	public int getEscapeMode() {
		return escapeMode;
	}
	
	public CodeWriter append(Identifier id) {
		append(id.toString());
		return this;
	}
	
	public CodeWriter append(String s) {
		indenter.appendTabsIfRequired(s);
		sb.append(s);
		return this;
	}

	public CodeWriter appendRaw(String s) {
		sb.append(s);
		return this;
	}

	public CodeWriter append(char c) {
		indenter.appendTabsIfRequired(c);
		sb.append(c);
		return this;
	}

	public CodeWriter trimLast(int count) {
		sb.setLength(sb.length()-count);
		return this;
	}

	public CodeWriter indent() {
		indenter.indent();
		return this;
	}

	public CodeWriter dedent() {
		indenter.dedent();
		return this;
	}

	public CodeWriter newLine() {
		append('\n');
		return this;
	}

	@Override
	public String toString() {
		return sb.toString();
	}

	public Dialect getDialect() {
		return dialect;
	}


	public boolean isGlobalContext() {
		return context.isGlobalsContext();
	}

	public CodeWriter newLocalWriter() {
		return new CodeWriter(dialect, context.newLocalContext(), escapeMode, sb, indenter);
	}

	public CodeWriter newChildWriter() {
		return newChildWriter(context.newChildContext());
	}

	public CodeWriter newChildWriter(Context context) {
		return new CodeWriter(dialect, context, escapeMode, sb, indenter);
	}

	public CodeWriter newMemberWriter() {
		Context context = this.context.newLocalContext();
		context.setParentContext(this.context);
		return new CodeWriter(dialect, context, escapeMode, sb, indenter);
	}

	public CodeWriter newInstanceWriter(CategoryType type) {
		return new CodeWriter(dialect, context.newInstanceContext(type, false), escapeMode, sb, indenter);
	}

	public CodeWriter newDocumentWriter() {
		return new CodeWriter(dialect, context.newDocumentContext(false), escapeMode, sb, indenter);
	}


}
